src/main/java/com/example/WMS/services/servicesimpl/IncomingGoodsServicesImpl.java

```diff
@@ -0,0 +1,98 @@
+package com.example.WMS.services.servicesimpl;
+
+import com.example.WMS.dtos.*;
+import com.example.WMS.entity.IncomingGoods;
+import com.example.WMS.entity.ProductDetails;
+import com.example.WMS.exceptions.OrderNotFound;
+import com.example.WMS.exceptions.ProductNotPresent;
+import com.example.WMS.handlers.ProductDetailsHandler;
+import com.example.WMS.handlers.IncomingGoodsHandler;
+import com.example.WMS.services.IncomingGoodsServices;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+
+@Service
+@Slf4j
+public class IncomingGoodsServicesImpl implements IncomingGoodsServices {
+
+    @Autowired
+    private IncomingGoodsHandler incomingGoodsHandler;
+
+    @Autowired
+    private ProductDetailsHandler productDetailsHandler;
+
+    public ValidationDto createIncomingGoods(AddStockDto addStockDto) {
+        String uniqueID = UUID.randomUUID().toString();
+        IncomingGoods incomingGoods = IncomingGoods.builder()
+                .incomingGoodsId(uniqueID)
+                .createdDate(new Date().getTime())
+                .merchantId(addStockDto.getMerchantId())
+                .productId(addStockDto.getProductId())
+                .quantity(addStockDto.getQuantity())
+                .build();
+        ProductDetails productDetails = productDetailsHandler.getProductDetails();
+        log.info("Product details received from file are: {}", productDetails);
+
+        try {
+            if(productDetails == null) {
+                throw new ProductNotPresent();
+            }
+            log.info("Product details received from file are: {}", productDetails);
+        } catch (ProductNotPresent e) {
+            log.error("Product is not present! {}", e);
+            return ValidationDto.builder()
+                    .isValid(false)
+                    .reason("product not found")
+                    .build();
+        } catch (Exception e) {
+            log.error("Some error has occurred: {}", e);
+            return ValidationDto.builder()
+                    .isValid(false)
+                    .reason("Some error occurred")
+                    .build();
+        }
+        incomingGoodsHandler.write(incomingGoods);
+        productDetails.setQuantity(productDetails.getQuantity() + incomingGoods.getQuantity());
```

Comment:
If 2 Incoming goods have been inputted at the same time
first is having quantity of 10
second is having quantity of 5
and the existing quantity is 4
expected is 19 but since both request are at same time it may not update the correct quantity both process will read get quantity as 4 and it may update to 14 or 9 depending on which one writes first

Reply:
I believe i have fixed this issue by putting the read and write functions for the product details in a synchronised block, so only 1 thread can read and write (edit) the product details at a time. Let me know if there is a more efficient way to do this or if this way does not work.

Reply:
Can you try writing one Junit test case to test this scenario ?

Reply:
Also, won't it increase the API response time? 
Try thinking of some other change which you can do to handle this scenario.